### PR TITLE
Refactor package structure

### PR DESCRIPTION
**Pull request by bleykauf**

### CODE REVIEW

**Comment by bleykauf** [comment]
Might be useful: https://python.tutorialink.com/run-makefile-on-pip-install/

**Comment by bleykauf** [comment]
Progress (partially copied from https://github.com/linien-org/linien/issues/275#issuecomment-1162753047):

* [x] building the FPGA image via build_fpga_image.sh
* [x] building standalone binaries of linien-gui with pyinstaller pyinstaller.spec on both Linux and Window
* [x] running linien-server on the RedPitaya and connecting to it
* [x] running `pytest .`: note that the `linien-*` packages need to be installed via pip first
* [x] uploading to testpypi
* [ ] github Actions
 * [ ] installation of the development version via the client/gui
* [ ] use .devX versions according to [PEP440](https://peps.python.org/pep-0440/) for development versions
* [ ] implement installation of server via pip

**Comment by doronbehar** [comment]
Thanks for the updates @bleykauf, I appreciate your work. Thanks to my small experience as a hobbyist in small scale devops prior to my bachelor's degree in physics, I noticed a few ways to improve the way the server runs:

The best practice way of running such a python server is via [Systemd](https://systemd.io/) which RedPitaya's OS (based on debian) is using as an init system. Before I knew this PR, I've written a service file which you might want to use:

```service
# linien.service
[Unit]
Description=Spectroscopy lock server for RedPitaya
Wants=network-online.target
After=network-online.target

[Service]
# Change this value either with `systemctl edit linien-server.service`, run
# `man systemctl(1)` to read documentation for this command. Perhaps this
# should be read from a file such as `/opt/linien/port` instead.
Environment="LINIEN_PORT=8000"
ExecStart=/usr/bin/env python3 /usr/lib/python3.10/site-packages/linien/server/server.py
# stop / start ethernet blinking using: mdio-tool[1], see:
# https://github.com/RedPitaya/RedPitaya/issues/205
#
# [1]: https://github.com/linien-org/mdio-tool
ExecStartPre=/usr/bin/mdio-tool w eth0 0x1b 0x0000
ExecStopPost=/usr/bin/mdio-tool w eth0 0x1b 0x0f00

[Install]
WantedBy=multi-user.target

# vim:ft=systemd
```

This should replace the bash scripts and you should be able to see if it's running via `systemctl status linien.service` and use `journalctl --unit=linien.service` to view the logs `server.py` prints. Such a change would require also rewriting parts of the README, and assume the user is capable of using putty or something to ssh to their RedPitaya and manage the installation there. It should also be possible to make the `linien.service` start on boot with `systemctl enable linien.service`.

I'm also not sure the client should install the requirements `pyrp3` and `mdio-tool` the way it does currently in `linien_install_requirements.sh`. I think the best way to handle `mdio-tool` would be to create a `.deb` package in a https://github.com/linien-org/mdio-tool . Similarly, `pyrp3` should become a pypi package and be specified in `linien-server/setup.py`.

I also noticed there's an authentication system based on environment variables in addition to ssh's authentication. This perhaps requires writing the hash to a file such as `/opt/linien/auth.txt` instead of using environment variables, because it's a bit hard to pass environment variables to systemd services - see the comment at the `linien.service` file. Perhaps also the port should be read from a file such as `/opt/linien/port`.

To summarize, I'd be glad to help in this field, if you are interested in such changes, as they will change the UX quite a bit. If you'd like to keep things as they are regarding the installation experience etc, that's OK, I guess they work for most users, I haven't had bad experience with them yet.

I plan to try out linien as a replacement for this application: https://marceluda.github.io/rp_lock-in_pid/ with which I locked a Rubidium LASER with.

**Comment by bleykauf** [comment]
Thanks for all your suggestions. As you have seen, I opened separate issues for them.

Those would definitely some good QoL improvements, some of them I already attempted in the past but never finished up. Any help hear would be greatly appreciated.

I also had another look at this PR (see also my unfortunately unsuccessful [attempt](https://github.com/linien-org/linien/pull/292) at getting setuptools-scm to work; its fixable but not today...). In any case, I think there is no need to fix all of the goals listed above in this PR. The only open issue is to at least adjust the github actions to at least create the windows executable with the new layout. The action for Linux is currently broken, see https://github.com/linien-org/linien/issues/263 (I have the feeling that this is a relatively easy fix, it is just necessary to elevate user privileges in the container). After that is done i would merge to the develop branch.

If you feel motivated to already work on something before that I would suggest you branch off of this branch since the changes to the package structure are significant.

**Comment by bleykauf** [comment]
github action for pyinstaller on Windows works again but still no luck with the Linux version, see #263.

**Comment by doronbehar** [comment]
Haven't tested it, but perhaps you can thankfully also edit `linien_install_requirements.sh` to accommodate for https://github.com/linien-org/pyrp3/pull/2 and https://github.com/linien-org/pyrp3#installation ?

```diff
diff --git c/linien/server/linien_install_requirements.sh i/linien/server/linien_install_requirements.sh
index dee6fa6..056d167 100644
--- c/linien/server/linien_install_requirements.sh
+++ i/linien/server/linien_install_requirements.sh
@@ -19,19 +19,7 @@ mkdir linien
 
 echo 'installing pyrp3...'
 # install pyrp3
-cd /tmp/linien
-git clone https://github.com/linien-org/pyrp3.git
-cd pyrp3
-python3 setup.py install
-
-echo 'building monitor...'
-# build monitor shared library
-cd monitor
-make
-cp libmonitor.so /usr/lib/
-
-cd /tmp
-rm -R /tmp/linien
+pip install pyrp3
 
 # https://github.com/RedPitaya/RedPitaya/issues/205
 echo 'building ethernet blinking fix'
@@ -40,4 +28,4 @@ cd mdio-tool
 cmake .
 make
 rm -f /usr/bin/mdio-tool
-mv mdio-tool /usr/bin
\ No newline at end of file
+mv mdio-tool /usr/bin
```

**Comment by bleykauf** [comment]
This is already incorporated into both the master branch: no more need to to call this at all in `linien_install_requirements.sh`, it is simply a requirement in the `setup.py` of `linien-server`.

However, I just relalized that in the develop branch there was still a merge conflict present. Thank you for bringing that to my attention :)
